File: 99-Continue/165/LeafSum.py
from collections import deque

class Node:
    def __init__(self,data:int):
        self.data = data
        self.right = self.left = None
    
class BinaryTree:
    def __init__(self):
        self.root = None
    
    def buildTree(self, arr:list[str])->Node:
        if len(arr)==0:
            return self.root
        
        self.root = Node(int(arr[0]))
        queue = deque([self.root])
        i=1
        while queue and i<len(arr):
            current = queue.popleft()
            if i<len(arr) and arr[i]!="null":
                current.left = Node(int(arr[i]))
                queue.append(current.left)
            i+=1
            if i<len(arr) and arr[i]!="null":
                current.right = Node(int(arr[i]))
                queue.append(current.right)
            i+=1
            
        return self.root
    
    def sumLeaf(self,root:Node)->int:
        if root is None:
            return 0
        
        if root.left is None and root.right is None:
            return root.data
        
        return self.sumLeaf(root.left)+self.sumLeaf(root.right)
        
def main():
    int(input())
    s = input().split()
    tree = BinaryTree()
    tree.buildTree(s)
    print(tree.sumLeaf(tree.root))

if __name__=="__main__":
    main()        
        
            
            
            
            
            

File: 99-Continue/165/LeafSum.java
import java.util.*;

public class LeafSum {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        sc.nextLine();
        String s = sc.nextLine();
        String[] strarr = s.split(" ");
        BinaryTree tree = new BinaryTree();
        tree.buildTree(strarr);
        System.out.println(tree.leafSum(tree.root));
        sc.close();
    }
}

class Node{
    int data;
    Node left,right;
    Node(int data){
        this.data = data;
        this.left = this.right = null;
    }
}

class BinaryTree{
    Node root;
    BinaryTree(){
        this.root=null;
    }
    
    Node buildTree(String[] arr){
        if(arr.length==0){
            return root;
        }
        
        root = new Node(Integer.parseInt(arr[0]));
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        int i=1;
        while(!queue.isEmpty() && i<arr.length){
            Node current = queue.poll();
            if(i<arr.length && !arr[i].equals("null")){
                current.left = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.left);
            }
            i++;
            if(i<arr.length && !arr[i].equals("null")){
                current.right = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.right);
            }
            i++;
        }
        return root;
    }
    
    int leafSum(Node root){
        if(root==null){
            return 0;
        }
        
        if(root.left==null && root.right==null){
            return root.data;
        }

        return leafSum(root.left)+leafSum(root.right);
    }
}